### DISCUSSION TITLE
Expose agent property of node-fetch

### DISCUSSION THREAD

**Discussion by TekSiDoT** [remix-run/remix#3530] [Q&A] [answered]
We're using a dev-machine local proxy to upgrade local dev-server non-HTTPS calls to HTTPS calls based on self-signed certificates. 

When using fetch() from within Remix, a `FetchError / _self signed certificate_` error is thrown.

The standard way to resolve this error in node-fetch would be to supply a custom `http.agent` with the `rejectUnauthorized` option set to `false`.  Unfortunately, the agent option is not part of the [TS lib.dom `RequestInit`](https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L1470) options definition exposed in Remix.

This is probably expected behavior, because afaics Remix just exposes the Web Fetch API, which of course has no concerns as changing its agent... any ideas?

**Comment by TekSiDoT** [comment, marked as answer]
After doing some more digging:
* Remix doesn't actually wrap node-fetch, instead, it uses a for of web-std-io as a source for its @remix-run/web-fetch package
* Setting the node env variable NODE_TLS_REJECT_UNAUTHORIZED circumvents the problem - at least for us it's good enough for local dev

**Comment by cliffordfajardo** [comment]
I recently found out that Remix does expose the `agent` property for `fetch` and `Response` APIs.
To get this to work, you need to import `fetch` or `Response` from the `@remix-run/node` package as opposed to using the global versions of `fetch` and `Response` which default to the browser versions, which dont expose the extra options.


### TLS Workaround -- Per HTTP/S Call

```ts
import { fetch, Request, json, LoaderArgs } from '@remix-run/node'; 
import https from 'https'; // OR import 'http' if your remix server is running on http

export async function loader({ params }:LoaderArgs) {
    // Example 1
    const response1 = await fetch('example-api.com/people',{
        agent: new https.Agent({ rejectUnauthorized: true, requestCert: false})
    })
    const data1 = await response1.json();


    // Example 2
    const response2 = new Request('example-api.com/people', {
        agent: new https.Agent({ rejectUnauthorized: true, requestCert: false})
    })
    const data2 = await response2.json();


    return json({});
  }
```

### TLS Workaround  -- Application Wide Setting
I don't claim that this is the best solution, but its a good way to get unblocked to continue building during development.
**You can pass the `NODE_TLS_REJECT_UNAUTHORIZED` environment variable into your node app during development like this:**

```ts
// Package.json
"scripts": {
    "dev": "NODE_TLS_REJECT_UNAUTHORIZED=0 remix dev",
},
```
or

```ts
// entry.server.ts
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
```




### References / Related
- https://github.com/remix-run/remix/discussions/4059
- https://github.com/remix-run/remix/discussions/4282
